compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeToTransaction.cpp

```diff
@@ -200,16 +200,34 @@ LogicalResult convertOp(AMDAIE::NpuAddressPatchOp op,
 }
 
 LogicalResult convertOp(AMDAIE::NpuDmaWaitOp op, TransactionBuilder &builder) {
-  for (Value token : op.getAsyncTokens()) {
-    auto pushToQueueOp =
-        dyn_cast_if_present<AMDAIE::NpuPushToQueueOp>(token.getDefiningOp());
+  // Batch DMA operations with the same row, channel, and direction into a
+  // single TCT sync operation, as long as they have consecutive columns.
+  SmallVector<std::pair<AMDAIE::NpuPushToQueueOp, uint32_t>> columnBatches;
+  for (Value asyncToken : op.getAsyncTokens()) {
+    auto pushToQueueOp = dyn_cast_if_present<AMDAIE::NpuPushToQueueOp>(
+        asyncToken.getDefiningOp());
     if (!pushToQueueOp) {
       return op.emitOpError()
-             << "should operate on an `amdaie.push_to_queue` op";
+             << "should operate on an `amdaie.push_to_queue` op async token";
     }
+    if (!columnBatches.empty()) {
+      auto &[lastPushOp, lastColNum] = columnBatches.back();
+      if (lastPushOp.getRow() == pushToQueueOp.getRow() &&
+          lastPushOp.getCol() + lastColNum == pushToQueueOp.getCol() &&
```

Comment:
Shouldn't you sort the tokens on column number for this to always result the largest batches? Suppose the tokens are provided as (all have same row, direction, channel):

```
token_0: col == 0
token_1: col == 2
token_2: col == 1
```
With the current logic, won't this result in two TCT syncs:

```
TCT sync 1: token_0, token_2
TCT sync 2: token_1
```

While if you sort on column first:
```
token_0: col == 0
token_2: col == 1
token_1: col == 2
```
This would result in one TCT sync?
```
TCT sync 1: token_0, token_2, token_1
```

Reply:
The sorting is actually done inside through this map.
https://github.com/nod-ai/iree-amd-aie/blob/022bdf813d38d9888893519af1c991ce9c2d3dd3/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp#L161-L162

In the current implementation, both batching and sorting are performed in `AMDAIEFoldDmaWaits`.  Later, in `AMDAIEControlCodeToTransaction`, the batching condition is rechecked (matching `row`, `channel`, and `direction`)  to ensure that no other passes have altered the operations in between. Sorting, however, is not repeated.

For debugging purposes, it might be better to enforce this as an assertion: a single `dma_wait` operation 
should not result in multiple `TCT_sync` operations. This would prevent potential confusion when debugging 
by maintaining a clear one-to-one correspondence between `dma_wait` and its resulting control operations.

Reply:
> In the current implementation, both batching and sorting are performed in AMDAIEFoldDmaWaits.

That's a different pass, so we shouldn't make that assumption and just sort again here.

> For debugging purposes, it might be better to turn this to an assertion: a single dma_wait operation
should not result in multiple TCT_sync operations. 

Why not? You could put wait for multiple tokens, which is what the current implementation supports.